File: asteroid.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

// SNode - узел стека
type SNode struct {
	data pair
	next *SNode
}

// Stack - структура стека
type Stack struct {
	head *SNode
}

// pair - замена std::pair<int, string>
type pair struct {
	first  int
	second string
}

// SPUSH - добавление элемента в стек
func SPUSH(list *Stack, data pair) {
	newNode := &SNode{data: data, next: list.head}
	list.head = newNode
}

// SPOP - удаление верхнего элемента стека
func SPOP(list *Stack) {
	if list.head == nil {
		return
	}
	nextNode := list.head.next
	list.head = nextNode
	// В Go сборщик мусора автоматически удалит память
}

// String - реализация вывода для pair (аналог operator<<)
func (p pair) String() string {
	return fmt.Sprintf("%d %s", p.first, p.second)
}

// SPRINT - вывод содержимого стека
func SPRINT(list *Stack) {
	current := list.head
	for current != nil {
		fmt.Println(current.data)
		current = current.next
	}
}

// ReverseStack - переворачивание стека
func ReverseStack(list *Stack) {
	reversed := &Stack{}
	current := list.head
	for current != nil {
		SPUSH(reversed, current.data)
		current = current.next
	}
	list.head = reversed.head
}

// AsteroidCycle - обработка столкновений астероидов
func AsteroidCycle(aster *Stack) {
	changed := false
	for {
		changed = false
		temp := &Stack{}

		current := aster.head
		for current != nil {
			if temp.head != nil &&
				temp.head.data.second == "right" &&
				current.data.second == "left" {

				// Обработка столкновения
				if temp.head.data.first == current.data.first {
					SPOP(temp)
				} else if temp.head.data.first > current.data.first {
					temp.head.data.first -= current.data.first
				} else {
					current.data.first -= temp.head.data.first
					SPOP(temp)
					SPUSH(temp, current.data)
				}
				changed = true
			} else {
				SPUSH(temp, current.data)
			}
			current = current.next
		}

		// Восстанавливаем исходный порядок
		aster.head = nil
		current = temp.head
		for current != nil {
			SPUSH(aster, current.data)
			current = current.next
		}

		if !changed {
			break
		}
	}
}

// asteroid - основная функция для работы с астероидами
func asteroid() {
	fmt.Println("Введите начальные данные в формате:")
	fmt.Println("<масса> <направление (l/r)> stop")

	stack := &Stack{}
	scanner := bufio.NewScanner(os.Stdin)
	
	if scanner.Scan() {
		input := scanner.Text()
		tokens := strings.Fields(input)
		
		mass := 0
		expectMass := true
		
		for _, token := range tokens {
			if token == "stop" {
				break
			}
			
			if expectMass {
				var err error
				mass, err = strconv.Atoi(token)
				if err != nil {
					fmt.Printf("Ошибка: масса должна быть числом, получено: %s\n", token)
					return
				}
				expectMass = false
			} else {
				var direction string
				if token == "l" || token == "r" {
					if token == "l" {
						direction = "left"
					} else {
						direction = "right"
					}
					ast := pair{mass, direction}
					SPUSH(stack, ast)
					expectMass = true
				} else {
					fmt.Printf("Ошибка: направление должно быть 'l' или 'r', получено: %s\n", token)
					return
				}
			}
		}
	}

	// Обрабатываем астероиды
	ReverseStack(stack)
	AsteroidCycle(stack)
	fmt.Println("Результат:")
	SPRINT(stack)
}

File: asteroid.cpp
#include <iostream>
#include  <sstream>
using namespace std;

class SNode{
public:
    pair<int, string> data;
    SNode* next;
    SNode(pair<int, string> dt, SNode* nx) : data(dt)
                                , next(nx){}
};

class Stack{
public:
    SNode* head;
    Stack(SNode* node) : head(node){}
    Stack(pair<int, string> data) : head(new SNode{data, nullptr}){}
    Stack() : head(nullptr){}
};

void SPUSH(Stack& list, pair<int, string> data){
    SNode* new_SNode = new SNode{data, list.head};
    list.head = new_SNode;
}

void SPOP(Stack& list){
    SNode* Address = list.head -> next;
    delete list.head;
    list.head = Address;
}


ostream& operator<<(ostream& ss, pair<int, string> data){
    ss << data.first << " " << data.second << " ";
    return ss;
}

void SPRINT(Stack& list){
    SNode* Address = list.head;
    while (Address != nullptr){
        cout << Address->data << endl;
        Address = Address->next;
    }
}

void ReverseStack(Stack& list) {
    Stack reversed;
    SNode* current = list.head;
    while (current != nullptr) {
        SPUSH(reversed, current->data);
        current = current->next;
    }
    list.head = reversed.head;
}

void AsteroidCycle(Stack& aster){
    bool changed;
    do {
        changed = false;
        Stack temp;
        
        SNode* current = aster.head;
        while (current != nullptr) {
            if (temp.head != nullptr && 
                temp.head->data.second == "right" && 
                current->data.second == "left") {
                
                // Обработка столкновения
                if (temp.head->data.first == current->data.first) {
                    SPOP(temp);
                } else if (temp.head->data.first > current->data.first) {
                    temp.head->data.first -= current->data.first;
                } else {
                    current->data.first -= temp.head->data.first;
                    SPOP(temp);
                    SPUSH(temp, current->data);
                }
                changed = true;
            } else {
                SPUSH(temp, current->data);
            }
            current = current->next;
        }
        
        // Восстанавливаем исходный порядок
        aster.head = nullptr;
        current = temp.head;
        while (current != nullptr) {
            SPUSH(aster, current->data);
            current = current->next;
        }
        
    } while (changed);
}

void asteroid(){
    cout << "Введите начальные данные в формате:\n<масса> <направление (l/r)> stop" << endl;
    
    Stack stack;
    string input;
    
    cin.ignore();
    getline(cin, input);
    
    stringstream ss(input);
    string token;
    int mass = 0;
    string direction;
    bool expectMass = true;
    
    while(ss >> token){
        if (token == "stop"){
            break;
        }
        
        if (expectMass){
            mass = stoi(token);
            expectMass = false;
        } else {
            if (token == "l" || token == "r") {
                direction = (token == "l" ? "left" : "right");
                pair<int, string> ast = {mass, direction};
                SPUSH(stack, ast);
                expectMass = true;
            } else {
                cout << "Ошибка: направление должно быть 'l' или 'r', получено: " << token << endl;
                return;
            }
        }
    }
    
    // Обрабатываем астероиды
    ReverseStack(stack);
    AsteroidCycle(stack);
    cout << "Результат:" << endl;
    SPRINT(stack);
}